### DISCUSSION TITLE
HELP WITH JAVA PLEASEEEEE!!

### DISCUSSION THREAD

**Discussion by whosVii** [community/community#124717] [Programming Help]
### Body

First of all, english isn't my first language.
Second, i don't know anything about java, but i'm trying to do a snake game (and if it works, it's gonna be a gift for boyfriend, who studies programming, anyway!), i did a lot of stuff on my code, and it was everything great, until a import some differents fonts, i got to import those text fonts for my code, but then, the "Game Over" screen didn't appear anymore, and i don't know what to do anymore!!!! Please, can someone help me??? I'm totally desperate, I've tried everything, I've researched, but I think the error is very specific!! And I have no one to ask

I don't think that the problem was the new text fonts, but something is happening to the game over screen... it was working before....

PS.: since i dont know ANYTHING about JAVA (this "beautiful" language), i did my code based on videos on youtube that i watched to learn at least the basic (spoiler: i couldnt learn, it doesnt matter how I tried), and the base of the game i got here on github, that i found from a youtube video (i tried to send the creator a message, but no answers) but it was the basic of the basic, I put a lot of things, so it must have a lot of "wrong things". ( i got a lot of help from ChatGPT -- what a shame) 

But that is it! I'm having a problem with the Game Over screen that doesnt appears anymore when I lose the game...

Here is my complete code (it will have a lot of things in Portugues-BR, but it's nothing important, just notes that i took thorught of the code):

package GameFrame;

	import java.awt.*;
	import java.awt.event.*;
	import javax.swing.*;
	import java.util.Random;
import java.io.File;
	import java.io.IOException;

	
	
	public class GamePanel extends JPanel implements ActionListener{

		private static final long serialVersionUID = 1L;
		private JButton restartButton;
		static int WIDTH = 600;
		static int HEIGHT = 600;
		static int UNIT_SIZE = 20;
		static final int NUMBER_OF_UNITS = (WIDTH * HEIGHT) / (UNIT_SIZE * UNIT_SIZE);
		static final int MAX_SCORE = 2; //limite de pontuação

		
		// hold x and y coordinates for body parts of the snake
		final int x[] = new int[NUMBER_OF_UNITS];
		final int y[] = new int[NUMBER_OF_UNITS];
		
		
		// tamanho da cobra no início
		int length = 5;
		int foodEaten;
		int foodX;
		int foodY;
		char direction = 'D';
		boolean running = false; //pro game over
		Random random;
		Timer timer;
		Font youngSerifFont;
		Font concertOneFont;
		Font postsenOneFont;
		Font jersey25Font;
		Font jostMediumFont;
		Font josefinsSemiBoldFont;
		Font artNuvoFont;
		Font teachersFont;
		
		
		GamePanel() {
			random = new Random();
			this.setPreferredSize(new Dimension(WIDTH-2, HEIGHT-2));
			this.setBackground(Color.DARK_GRAY);
			this.setFocusable(true);
			this.addKeyListener(new MyKeyAdapter());
			
			
			//fonte de texto importada. YOUNG SERIF 
			try {
	            youngSerifFont = Font.createFont(Font.TRUETYPE_FONT, new File("fontes/YoungSerif-Regular.ttf")).deriveFont(Font.PLAIN, 25);
	            GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
	            ge.registerFont(youngSerifFont);
	        } catch (IOException | FontFormatException e) {
	            e.printStackTrace();
	            youngSerifFont = new Font("SansSerif", Font.PLAIN, 25); // Fallback para uma fonte padrão
	        }
			
			// fonte de texto importada. CONCERT ONE
			try {
				concertOneFont = Font.createFont(Font.TRUETYPE_FONT, new File("fontes/ConcertOne-Regular.ttf")).deriveFont(Font.PLAIN, 25);
	            GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
	            ge.registerFont(concertOneFont);
	        } catch (IOException | FontFormatException e) {
	            e.printStackTrace();
	            concertOneFont = new Font("SansSerif", Font.PLAIN, 25); // Fallback para uma fonte padrão
	        }
			
			// fonte de texto importada. POETSEN ONE
			try {
				postsenOneFont = Font.createFont(Font.TRUETYPE_FONT, new File("fontes/PoetsenOne-Regular.ttf")).deriveFont(Font.PLAIN, 25);
	            GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
	            ge.registerFont(postsenOneFont);
	        } catch (IOException | FontFormatException e) {
	            e.printStackTrace();
	            postsenOneFont = new Font("SansSerif", Font.PLAIN, 25); // Fallback para uma fonte padrão
	        }
			
			
			// fonte de texto importada. Jersey25
			try {
				jersey25Font = Font.createFont(Font.TRUETYPE_FONT, new File("fontes/Jersey25-Regular.ttf")).deriveFont(Font.PLAIN, 25);
				GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
				ge.registerFont(jersey25Font);
			} catch (IOException | FontFormatException e) {
				e.printStackTrace();
				jersey25Font = new Font("SansSerif", Font.PLAIN, 25); // Fallback para uma fonte padrão
				        }
			
		
		
						
			// fonte de texto importada. JosefinSans-SemiBold
						try {
							josefinsSemiBoldFont = Font.createFont(Font.TRUETYPE_FONT, new File("fontes/JosefinSans-SemiBold.ttf")).deriveFont(Font.PLAIN, 14);
							GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
							ge.registerFont(josefinsSemiBoldFont);
						} catch (IOException | FontFormatException e) {
							e.printStackTrace();
							josefinsSemiBoldFont = new Font("SansSerif", Font.PLAIN, 25); // Fallback para uma fonte padrão
							        }
			
			
							        
			
			// fonte de texto importada. Jost Medium
			try {
				jostMediumFont = Font.createFont(Font.TRUETYPE_FONT, new File("fontes/Jost-Medium.ttf")).deriveFont(Font.PLAIN, 25);
				GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
				ge.registerFont(jostMediumFont);
			} catch (IOException | FontFormatException e) {
				e.printStackTrace();
				jostMediumFont = new Font("SansSerif", Font.PLAIN, 25); // Fallback para uma fonte padrão
										        }
						
			// fonte de texto importada. ART NUVO -----> game over/you win
			try {
				artNuvoFont = Font.createFont(Font.TRUETYPE_FONT, new File("fontes/Art-nuvo.otf")).deriveFont(Font.PLAIN, 25);
				GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
				ge.registerFont(artNuvoFont);
			} catch (IOException | FontFormatException e) {
				e.printStackTrace();
				artNuvoFont = new Font("SansSerif", Font.PLAIN, 25); // Fallback para uma fonte padrão
													        }	
			
			// fonte de texto importada. TEACHERS ------> pro SCORE e RESTART // testar primeiro
						try {
							teachersFont = Font.createFont(Font.TRUETYPE_FONT, new File("fontes/Teachers-VariableFont_wght.ttf")).deriveFont(Font.PLAIN, 25);
							GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
							ge.registerFont(teachersFont);
						} catch (IOException | FontFormatException e) {
							e.printStackTrace();
							teachersFont = new Font("SansSerif", Font.PLAIN, 25); // Fallback para uma fonte padrão
																        }
			
						
			restartButton = new JButton("RESTART");
			restartButton.setFont(teachersFont.deriveFont(Font.PLAIN, 15)); // MUDAR FONTE (pequisar uma q combine com concertOne
			restartButton.setForeground(Color.WHITE);	
			Color buttonColor = new Color(50, 150, 255); // Cor azul claro
			restartButton.setBackground(buttonColor);

			
			
		    restartButton.addActionListener(new ActionListener() {
		       public void actionPerformed(ActionEvent e) {
		             restartGame();
		             requestFocusInWindow();
		            }

		    });
		    
		    restartButton.setBounds(250, 550, 100, 30);
		    setLayout(null);
		    add(restartButton);
		   
			play();
		}
		
		public void play() {
			addFood();
			running = true;
		
			if (timer == null) { //pra manter a mesma velocidade do primeiro jogo nos outros
			timer = new Timer(80, this);  //velocidade (se ficar fácil, aumentr
		}
			timer.start();	
		}
		
	    public void restartGame() { //pra reiniciar o jogo funcionando
	        timer.stop();
	    	length = 5;
	        foodEaten = 0;
	        direction = 'D';
	        running = true;
	        // pra resetar a posição da cobra
	        for (int i = 0; i < length; i++) {
	            x[i] = 0;
	            y[i] = 0;
	       
	        }
	        
	          
	        // isso é pra permanecer o tamanho da tela e da cobra depois que reinicia o jogo
	        int WIDTH = 600;
			int HEIGHT = 600;
			int UNIT_SIZE = 20;
	        UNIT_SIZE = WIDTH / 20;
	       
	
	        restartButton.setBounds(250, 550, 100, 30);     
	        play();
	        repaint();
	    }
	    
		

		@Override
		public void paintComponent(Graphics graphics) {
			super.paintComponent(graphics);
			draw(graphics);
		}
		
		public void move() {
			for (int i = length; i > 0; i--) {
				// shift the snake one unit to the desired direction to create a move
				x[i] = x[i-1];
				y[i] = y[i-1];
			}
			
			if (direction == 'L') {
				x[0] = x[0] - UNIT_SIZE;
			} else if (direction == 'R') {
				x[0] = x[0] + UNIT_SIZE;
			} else if (direction == 'U') {
				y[0] = y[0] - UNIT_SIZE;
			} else {
				y[0] = y[0] + UNIT_SIZE;
			}	
		}
		
		public void checkFood() {
			if(x[0] == foodX && y[0] == foodY) {
				length++;
				foodEaten++;
				
				if(foodEaten >= MAX_SCORE) { //para o jogo qnd atinge 15 pontos
					running = false;
					timer.stop();
					repaint();
					
				}  else {
					addFood();
				}
			}
		}
		
		public void draw(Graphics graphics) {
			
			if (running) {
				graphics.setColor(new Color(255, 115, 90));
				graphics.fillOval(foodX, foodY, UNIT_SIZE, UNIT_SIZE);
				
				graphics.setColor(Color.white);
				graphics.fillRect(x[0], y[0], UNIT_SIZE, UNIT_SIZE);
				
				for (int i = 1; i < length; i++) {
					graphics.setColor(new Color(40, 200, 150));
					graphics.fillRect(x[i], y[i], UNIT_SIZE, UNIT_SIZE);
				}
	
				
				graphics.setColor(Color.white);
				graphics.setFont(teachersFont);
				FontMetrics metrics = getFontMetrics(graphics.getFont());
				graphics.drawString("SCORE: " + foodEaten, (WIDTH - metrics.stringWidth("SCORE: " + foodEaten)) / 2, graphics.getFont().getSize());
			
			} else {
				if (foodEaten >=MAX_SCORE) {
					
					gameWon(graphics);
					textinhoP(graphics); // EU CONSEGUI SOZINHA COLOCAR MAIS UM TEXTO!!!!!!!!!!!!!!!!
				
				
			} else {
				gameOver(graphics);
			}
		}
	}
		
		public void addFood() {
			foodX = random.nextInt((int)(WIDTH / UNIT_SIZE))*UNIT_SIZE;
			foodY = random.nextInt((int)(HEIGHT / UNIT_SIZE))*UNIT_SIZE;
		}
		
		public void checkHit() {
			// check if head run into its body
			for (int i = length; i > 0; i--) {
				if (x[0] == x[i] && y[0] == y[i]) {
					running = false;
					break;
				}
			}
			
			// check if head run into walls
			if (x[0] < 0 || x[0] > WIDTH || y[0] < 0 || y[0] > HEIGHT) {
				running = false;
				
			}
			
			if(!running) {
				timer.stop();
				 System.out.println("Game Over - Collision detected"); // Log para depuração
				repaint();
			}
		}
		
		//gráficos de quando perde (game over)
		public void gameOver(Graphics graphics) {
			graphics.setColor(new Color(60, 230, 452));
			graphics.setFont(artNuvoFont.deriveFont(Font.PLAIN, 50));
			FontMetrics metrics = getFontMetrics(graphics.getFont());
			graphics.drawString("Game Over", (WIDTH - metrics.stringWidth("Game Over")) / 2, HEIGHT / 2);
			
			graphics.setColor(Color.white);
			graphics.setFont(teachersFont.deriveFont(Font.BOLD, 25));
			metrics = getFontMetrics(graphics.getFont());
			graphics.drawString("SCORE: " + foodEaten, (WIDTH - metrics.stringWidth("SCORE: " + foodEaten)) / 2, graphics.getFont().getSize());

		
		}
		
		
		//gráficos de quando ganha
		public void gameWon(Graphics graphics) {
	        graphics.setColor(Color.green);
	        graphics.setFont(artNuvoFont.deriveFont(Font.PLAIN, 90));	
	        FontMetrics metrics = getFontMetrics(graphics.getFont());
	        graphics.drawString("You Win!", (WIDTH - metrics.stringWidth("You Win!")) / 2, HEIGHT / 2);
	        

	        

	        graphics.setColor(Color.white);
	        graphics.setFont(teachersFont.deriveFont(Font.BOLD, 25));
	        metrics = getFontMetrics(graphics.getFont());
	        graphics.drawString("SCORE: " + foodEaten, (WIDTH - metrics.stringWidth("SCORE: " + foodEaten)) / 2,
	                graphics.getFont().getSize());
	    }
		
		//graficos do textinho
		public void textinhoP(Graphics graphics) {
			graphics.setColor(Color.pink);
	        graphics.setFont(artNuvoFont.deriveFont(Font.PLAIN, 20));	
	        FontMetrics metrics = getFontMetrics(graphics.getFont());
	        graphics.drawString("15 pontos = 15/06", (WIDTH - metrics.stringWidth("15 pontos = 15/06"))/2, HEIGHT/2);
			
		}
		
		
		@Override
		public void actionPerformed(ActionEvent arg0) {
			try {
			if (running) {
				move();
				checkFood();
				checkHit();
			}
			repaint();
		} catch (Exception e) {
            e.printStackTrace();
            System.err.println("Error in actionPerformed: " + e.getMessage());
        }
		}
		
		public class MyKeyAdapter extends KeyAdapter {
			@Override
			public void keyPressed(KeyEvent e) {
				switch(e.getKeyCode()) {
					case KeyEvent.VK_LEFT:
						if (direction != 'R') {
							direction = 'L';
						}
						break;
						
					case KeyEvent.VK_A:
						if (direction != 'R') {
							direction = 'L';
						}
						break;
						
						
					case KeyEvent.VK_RIGHT:
						if (direction != 'L') {
							direction = 'R';
						}
						break;
						
					case KeyEvent.VK_D:
						if (direction != 'L') {
							direction = 'R';
						}
						break;
			
					case KeyEvent.VK_UP:
						if (direction != 'D') {
							direction = 'U';
							
						}
						break;
						
					
					case KeyEvent.VK_W:
						if (direction != 'D') {
							direction = 'U';
							
				
						}
						break;
						
					case KeyEvent.VK_DOWN:
						if (direction != 'U') {
							direction = 'D';
						}
						break;
						
						
					case KeyEvent.VK_S:
						if (direction != 'U') {
							direction = 'D';
						}
						break;	
						
						
				}
			}
		}
	}

What appears to me, when I lose in the game, is this message: "at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)" (what is this?????????)

The game stops, i can click the button to restart and i restart the game, but, the only thing wrong is the game over screen that doesn't appear!!!!!!!! 

The code has another two public classes:

package GameFrame;

import javax.swing.JFrame;

public class GameFrame extends JFrame{

	private static final long serialVersionUID = 1L;

	GameFrame() {
		GamePanel panel = new GamePanel();
		this.add(panel);
		this.setTitle("snake"); //título do jogo
		this.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		this.setResizable(false);
		this.pack();
		this.setVisible(true);
		this.setLocationRelativeTo(null);
		

	}
}

And

package GameFrame;

public class SnakeGame {

		public static void main(String[] args) {
			new GameFrame();
		}

	}

HELP ME PLEASE

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by PushpitSB** [comment]
To help you fix the issue where the "Game Over" screen does not appear, let's follow these steps to debug and correct your code. I'll make sure to add comments and explain each part of the process.

### Analysis and Debugging
1. **Ensure the `gameOver` Method is Called:**
   - The `gameOver` method should be called when the game is not running and the snake collides with itself or the walls.

2. **Check the `draw` Method:**
   - Verify that the `draw` method calls `gameOver` correctly when `running` is false.

### Modifications and Fixes

I'll highlight any necessary changes and ensure the game displays the "Game Over" screen correctly.

#### Updated `GamePanel` Class

```java
package GameFrame;

import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.util.Random;
import java.io.File;
import java.io.IOException;

public class GamePanel extends JPanel implements ActionListener {

	private static final long serialVersionUID = 1L;
	private JButton restartButton;
	static int WIDTH = 600;
	static int HEIGHT = 600;
	static int UNIT_SIZE = 20;
	static final int NUMBER_OF_UNITS = (WIDTH * HEIGHT) / (UNIT_SIZE * UNIT_SIZE);
	static final int MAX_SCORE = 2; // limite de pontuação

	// hold x and y coordinates for body parts of the snake
	final int x[] = new int[NUMBER_OF_UNITS];
	final int y[] = new int[NUMBER_OF_UNITS];

	// tamanho da cobra no início
	int length = 5;
	int foodEaten;
	int foodX;
	int foodY;
	char direction = 'D';
	boolean running = false; // pro game over
	Random random;
	Timer timer;
	Font youngSerifFont;
	Font concertOneFont;
	Font postsenOneFont;
	Font jersey25Font;
	Font jostMediumFont;
	Font josefinsSemiBoldFont;
	Font artNuvoFont;
	Font teachersFont;

	GamePanel() {
		random = new Random();
		this.setPreferredSize(new Dimension(WIDTH - 2, HEIGHT - 2));
		this.setBackground(Color.DARK_GRAY);
		this.setFocusable(true);
		this.addKeyListener(new MyKeyAdapter());

		// Load all custom fonts
		loadFonts();

		restartButton = new JButton("RESTART");
		restartButton.setFont(teachersFont.deriveFont(Font.PLAIN, 15)); // MUDAR FONTE
		restartButton.setForeground(Color.WHITE);
		Color buttonColor = new Color(50, 150, 255); // Cor azul claro
		restartButton.setBackground(buttonColor);

		restartButton.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				restartGame();
				requestFocusInWindow();
			}
		});

		restartButton.setBounds(250, 550, 100, 30);
		setLayout(null);
		add(restartButton);

		play();
	}

	private void loadFonts() {
		try {
			youngSerifFont = Font.createFont(Font.TRUETYPE_FONT, new File("fontes/YoungSerif-Regular.ttf")).deriveFont(Font.PLAIN, 25);
			GraphicsEnvironment.getLocalGraphicsEnvironment().registerFont(youngSerifFont);
		} catch (IOException | FontFormatException e) {
			e.printStackTrace();
			youngSerifFont = new Font("SansSerif", Font.PLAIN, 25);
		}

		// Repeat the above for all other fonts: concertOneFont, postsenOneFont, etc.
		// This will keep the constructor cleaner and focused
	}

	public void play() {
		addFood();
		running = true;
		if (timer == null) { // Maintain the same speed for subsequent games
			timer = new Timer(80, this);
		}
		timer.start();
	}

	public void restartGame() {
		timer.stop();
		length = 5;
		foodEaten = 0;
		direction = 'D';
		running = true;

		for (int i = 0; i < length; i++) {
			x[i] = 0;
			y[i] = 0;
		}

		restartButton.setBounds(250, 550, 100, 30);
		play();
		repaint();
	}

	@Override
	public void paintComponent(Graphics graphics) {
		super.paintComponent(graphics);
		draw(graphics);
	}

	public void move() {
		for (int i = length; i > 0; i--) {
			x[i] = x[i - 1];
			y[i] = y[i - 1];
		}

		switch (direction) {
			case 'L':
				x[0] -= UNIT_SIZE;
				break;
			case 'R':
				x[0] += UNIT_SIZE;
				break;
			case 'U':
				y[0] -= UNIT_SIZE;
				break;
			case 'D':
				y[0] += UNIT_SIZE;
				break;
		}
	}

	public void checkFood() {
		if (x[0] == foodX && y[0] == foodY) {
			length++;
			foodEaten++;
			if (foodEaten >= MAX_SCORE) {
				running = false;
				timer.stop();
				repaint();
			} else {
				addFood();
			}
		}
	}

	public void draw(Graphics graphics) {
		if (running) {
			graphics.setColor(new Color(255, 115, 90));
			graphics.fillOval(foodX, foodY, UNIT_SIZE, UNIT_SIZE);

			graphics.setColor(Color.white);
			graphics.fillRect(x[0], y[0], UNIT_SIZE, UNIT_SIZE);

			for (int i = 1; i < length; i++) {
				graphics.setColor(new Color(40, 200, 150));
				graphics.fillRect(x[i], y[i], UNIT_SIZE, UNIT_SIZE);
			}

			graphics.setColor(Color.white);
			graphics.setFont(teachersFont);
			FontMetrics metrics = getFontMetrics(graphics.getFont());
			graphics.drawString("SCORE: " + foodEaten, (WIDTH - metrics.stringWidth("SCORE: " + foodEaten)) / 2, graphics.getFont().getSize());
		} else {
			if (foodEaten >= MAX_SCORE) {
				gameWon(graphics);
				textinhoP(graphics);
			} else {
				gameOver(graphics);
			}
		}
	}

	public void addFood() {
		foodX = random.nextInt((int) (WIDTH / UNIT_SIZE)) * UNIT_SIZE;
		foodY = random.nextInt((int) (HEIGHT / UNIT_SIZE)) * UNIT_SIZE;
	}

	public void checkHit() {
		for (int i = length; i > 0; i--) {
			if (x[0] == x[i] && y[0] == y[i]) {
				running = false;
				break;
			}
		}

		if (x[0] < 0 || x[0] > WIDTH || y[0] < 0 || y[0] > HEIGHT) {
			running = false;
		}

		if (!running) {
			timer.stop();
			System.out.println("Game Over - Collision detected");
			repaint();
		}
	}

	public void gameOver(Graphics graphics) {
		graphics.setColor(new Color(60, 230, 452));
		graphics.setFont(artNuvoFont.deriveFont(Font.PLAIN, 50));
		FontMetrics metrics = getFontMetrics(graphics.getFont());
		graphics.drawString("Game Over", (WIDTH - metrics.stringWidth("Game Over")) / 2, HEIGHT / 2);

		graphics.setColor(Color.white);
		graphics.setFont(teachersFont.deriveFont(Font.BOLD, 25));
		metrics = getFontMetrics(graphics.getFont());
		graphics.drawString("SCORE: " + foodEaten, (WIDTH - metrics.stringWidth("SCORE: " + foodEaten)) / 2, graphics.getFont().getSize());
	}

	public void gameWon(Graphics graphics) {
		graphics.setColor(Color.green);
		graphics.setFont(artNuvoFont.deriveFont(Font.PLAIN, 90));
		FontMetrics metrics = getFontMetrics(graphics.getFont());
		graphics.drawString("You Win!", (WIDTH - metrics.stringWidth("You Win!")) / 2, HEIGHT / 2);

		graphics.setColor(Color.white);
		graphics.setFont(teachersFont.deriveFont(Font.BOLD, 25));
		metrics = getFontMetrics(graphics.getFont());
		graphics.drawString("SCORE: " + foodEaten, (WIDTH - metrics.stringWidth("SCORE: " + foodEaten)) / 2, graphics.getFont().getSize());
	}

	public void textinhoP(Graphics graphics) {
		graphics.setColor(Color.pink);
		graphics.setFont(artNuvoFont.deriveFont(Font.PLAIN, 20));
		FontMetrics metrics = getFontMetrics(graphics.getFont());
		graphics.drawString("15 pontos = 15/06", (WIDTH - metrics.stringWidth("15 pontos = 15/06")) / 2, HEIGHT / 2 + 30);
	}

	@Override
	public void actionPerformed(ActionEvent arg0) {
		try {
			if (running) {
				move();
				checkFood();
				checkHit();
			}
			repaint();
		} catch (Exception e) {


			e.printStackTrace();
			System.err.println("Error in actionPerformed: " + e.getMessage());
		}
	}

	public class MyKeyAdapter extends KeyAdapter {
		@Override
		public void keyPressed(KeyEvent e) {
			switch (e.getKeyCode()) {
				case KeyEvent.VK_LEFT:
				case KeyEvent.VK_A:
					if (direction != 'R') {
						direction = 'L';
					}
					break;
				case KeyEvent.VK_RIGHT:
				case KeyEvent.VK_D:
					if (direction != 'L') {
						direction = 'R';
					}
					break;
				case KeyEvent.VK_UP:
				case KeyEvent.VK_W:
					if (direction != 'D') {
						direction = 'U';
					}
					break;
				case KeyEvent.VK_DOWN:
				case KeyEvent.VK_S:
					if (direction != 'U') {
						direction = 'D';
					}
					break;
			}
		}
	}
}
```

### Additional Classes

```java
package GameFrame;

import javax.swing.JFrame;

public class GameFrame extends JFrame {

	private static final long serialVersionUID = 1L;

	GameFrame() {
		GamePanel panel = new GamePanel();
		this.add(panel);
		this.setTitle("Snake");
		this.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		this.setResizable(false);
		this.pack();
		this.setVisible(true);
		this.setLocationRelativeTo(null);
	}
}
```

```java
package GameFrame;

public class SnakeGame {

	public static void main(String[] args) {
		new GameFrame();
	}
}
```

### Explanation and Debugging Tips

1. **Loading Fonts:**
   - The method `loadFonts` loads all the custom fonts. This helps keep the constructor clean.

2. **Restart Game:**
   - Ensures the `restartGame` method properly resets all game variables and state.

3. **Game Over Screen:**
   - Check if `gameOver` is correctly called within the `draw` method when `running` is false.

4. **Debugging:**
   - Added a `System.out.println("Game Over - Collision detected");` statement to debug when a collision is detected.

By following these steps and reviewing the modifications, you should be able to troubleshoot and correct the issue where the "Game Over" screen does not appear.
